spec.html

```diff
@@ -9049,17 +9049,22 @@ <h1>[[Construct]] ( _argumentsList_, _newTarget_ )</h1>
     </emu-clause>
 
     <emu-clause id="sec-createbuiltinfunction" aoid="CreateBuiltinFunction">
-      <h1>CreateBuiltinFunction ( _steps_, _internalSlotsList_ [ , _realm_ [ , _prototype_ ] ] )</h1>
-      <p>The abstract operation CreateBuiltinFunction takes arguments _steps_ and _internalSlotsList_ (a List of names of internal slots) and optional arguments _realm_ and _prototype_. _internalSlotsList_ contains the names of additional internal slots that must be defined as part of the object. This operation creates a built-in function object. It performs the following steps when called:</p>
+      <h1>CreateBuiltinFunction ( _steps_, _additionalInternalSlotsList_ [ , _realm_ [ , _prototype_ ] ] )</h1>
+      <p>The abstract operation CreateBuiltinFunction takes arguments _steps_ and optional arguments _additionalInternalSlotsList_ (a List of internal slot names), _realm_, and _prototype_. _additionalInternalSlotsList_ contains the names of additional internal slots that must be defined as part of the object. This operation creates a built-in function object. It performs the following steps when called:</p>
       <emu-alg>
         1. Assert: _steps_ is either a set of algorithm steps or other definition of a function's behaviour provided in this specification.
         1. If _realm_ is not present, set _realm_ to the current Realm Record.
         1. Assert: _realm_ is a Realm Record.
         1. If _prototype_ is not present, set _prototype_ to _realm_.[[Intrinsics]].[[%Function.prototype%]].
-        1. Let _func_ be a new built-in function object that when called performs the action described by _steps_. The new function object has internal slots whose names are the elements of _internalSlotsList_, and an [[InitialName]] internal slot.
-        1. Set _func_.[[Realm]] to _realm_.
-        1. Set _func_.[[Prototype]] to _prototype_.
+        1. If the built-in function object is implemented using ECMAScript code, then
```

Comment:
I am against this change. Switching on implementation details in normative prose seems bad to me.

Reply:
In the current spec, I think the only obvious example of switching on an implementation detail is:
- `If an implementation-defined debugging facility is available and enabled, ...`.

For "host", there's:
- `If the host requires use of an exotic object to serve as realm's global object, ...`
- `If the host requires that the *this* binding in realm's global scope return an object other than the global object, ...`
- `If <various conditions> and ! HostHasSourceTextAvailable(_func_) is *true*, ...`

That's all I could find.

Reply:
@bakkot:
> Switching on implementation details in normative prose seems bad to me.

[9.3 Built-in Function Objects](https://www.ecma-international.org/ecma-262/#sec-built-in-function-objects) says that *if* you implement a built-in function as an exotic function, then it must satisfy some extra requirements (that run-of-the-mill exotic functions don't have to). So that is normative prose that is switching on an implementation 'detail'.  The proposed text that you were objecting to seems comparable, just in algorithmic form.